### PR TITLE
[WIP] Normalize NANs and Zeroes (JNI bindings)

### PR DESCRIPTION
**Pull request by mythrocks**
4033cf8 added `cudf::normalize_nans_and_zeros`. This commit
adds the JNI bindings, for use from Spark.

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by jlowe** [comment]
add to whitelist

**Comment by jlowe** [comment]
rerun tests

**Comment by raydouglass** [comment]
rerun tests

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 1]
```diff
@@ -457,6 +459,30 @@ JNIEXPORT jobject JNICALL Java_ai_rapids_cudf_ColumnVector_exactQuantile(JNIEnv
...
+  JNI_NULL_CHECK(env, input_column, "native handle is null", 0);
+  try {
+    std::unique_ptr<column> p_normalized_column
+        = cudf::normalize_nans_and_zeros(
+            gdf_column_to_view(*reinterpret_cast<gdf_column*>(input_column)));
+
+    auto ret = reinterpret_cast<jlong>(
+        new gdf_column(view_to_gdf_column(p_normalized_column->mutable_view())));
+
+    p_normalized_column->release();
```
Comment (jrhemstad) [inline java/src/main/native/src/ColumnVectorJni.cpp]: This calls release without doing anything with the return value, so this effectively deletes the contents of the column. Your `ret` will have dangling pointers to free'd memory.

**Comment by jrhemstad** [comment]
By the way, the APIs in `interop.hpp` are untested and will definitely break for some types.

File: java/src/main/native/src/ColumnVectorJni.cpp (hunk 1 above)
Comment (mythrocks) [reply java/src/main/native/src/ColumnVectorJni.cpp]: Ok, I think I misunderstood the semantics of `column::release()`. Let me try again.

**Comment by raydouglass** [comment]
rerun tests

File: java/src/main/native/src/ColumnVectorJni.cpp (hunk 1 above)
Comment (mythrocks) [reply java/src/main/native/src/ColumnVectorJni.cpp]: We're in something of a bind. :/ 
One tentative solution to keeping the `column::contents` alive (long enough for `gdf_column` to make it back to Java-land) might be:
```c++
    auto ret = reinterpret_cast<jlong>(
        new gdf_column(view_to_gdf_column(p_normalized_column->mutable_view())));

    auto contents = p_normalized_column->release();
    (void)contents.data.release();      // Data-buffer contents have been adopted already
    (void)contents.null_mask.release(); // by the gdf_column being returned. Relinquish ownership.

    return ret;
```
But this would leave the `device_buffer` object itself on the heap. :/

Comment (mythrocks) [reply java/src/main/native/src/ColumnVectorJni.cpp]: Chaps, I was wondering if this version is more acceptable. I realize that we might leak the `device_buffer` members (the pointers, not the memory it points to). I don't see a way around it without `DeviceVector` switching over to use `column` instead of translating from `gdf_column`.

**Comment by mythrocks** [comment]
> By the way, the APIs in `interop.hpp` are untested and will definitely break for some types.

ACK. Thank you for pointing this out. I suspect this is acceptable until `ColumnVector` and the JNI code switches to `cudf++` `column`s.

File: java/src/main/native/src/ColumnVectorJni.cpp (hunk 1 above)
Comment (jrhemstad) [reply java/src/main/native/src/ColumnVectorJni.cpp]: I don't see how you're going to be able to use this libcudf++ API until the JNI machinery has been updated to use the new `cudf::column` and associated types.

Comment (mythrocks) [reply java/src/main/native/src/ColumnVectorJni.cpp]: >  until the JNI machinery has been updated to use the new `cudf::column` and associated types.

I agree. The Gordian Knot tightens. :/
Would you (and @jlowe, and @revans2) be averse to this current version being checked in (to work with `gdf_column`) until we sort out our move to `cudf::column`?

Comment (jrhemstad) [reply java/src/main/native/src/ColumnVectorJni.cpp]: That's purely up to you guys.

**Comment by jlowe** [comment]
I'd rather not checkin code that's known to leak memory.  This code needs to be rewritten anyway when the JNI moves to tracking cudf::column rather than gdf_column, and I don't think this is critical functionality before the JNI move needs to be completed.

Therefore I recommend marking this PR as WIP until the JNI layer switches to tracking cudf::column.

**Comment by harrism** [comment]
Moving to 0.12

**Comment by harrism** [comment]
@mythrocks @jlowe is this dead? Moving to 0.13 for now.

**Comment by jlowe** [comment]
@harrism No this is not dead, just taking a bit of an extended nap.  It was put on hold since we realized solving this without leaking requires the JNI move to the new cudf++ API.  We hope to have a PR posted where we do just that in a week or so.  Pushing this to 0.13 in the interim is fine.

**Comment by mythrocks** [comment]
Right, I'll get this in gear once we check in the JNI-side's move to `cudf++` APIs. 
(Note to self: Get this checked in, internally.)

**Comment by harrism** [comment]
Moving to 0.14.

**Comment by razajafri** [comment]
@mythrocks is this still valid?

**Comment by mythrocks** [comment]
Closing this, in favour of #4750.
